src/protocol/validation.rs

```diff
@@ -0,0 +1,234 @@
+    ///
+    /// * `provider` - The RPC provider to use for on-chain queries
+    /// * `block_id` - The block number to query at
+    /// * `component_id` - The component/pool address
+    /// * `tokens` - The tokens in the pool
+    ///
+    /// # Returns
+    ///
+    /// Returns `Ok(true)` if validation passes, `Ok(false)` if validation fails,
+    /// or `Err` if there was an error during validation
+    async fn validate(
+        &self,
+        provider: Arc<RootProvider<Ethereum>>,
+        block_id: u64,
+        component_id: &Bytes,
+        tokens: &[Token],
+    ) -> Result<bool, Box<dyn std::error::Error + Send + Sync>>;
+}
+
+/// Helper function to validate a ProtocolSim state if it implements Validator
+///
+/// This function attempts to cast the ProtocolSim to known types that implement Validator.
+/// If the type implements Validator, it validates and returns the result.
+/// If the type doesn't implement Validator, it returns None.
+pub async fn try_validate(
+    state: &dyn ProtocolSim,
+    provider: Arc<RootProvider<Ethereum>>,
+    block_id: u64,
+    component_id: &Bytes,
+    tokens: &[Token],
+) -> Option<Result<bool, Box<dyn std::error::Error + Send + Sync>>> {
+    // Try to cast to types that implement Validator
+    if let Some(uniswap_v2) = state
+        .as_any()
+        .downcast_ref::<UniswapV2State>()
+    {
+        return Some(
+            uniswap_v2
+                .validate(provider, block_id, component_id, tokens)
+                .await,
+        );
+    }
+
+    // Add more types here as they implement custom validation
+    // For now, all other types return None (no validation)
+
+    None
+}
+
+#[async_trait]
+impl Validator for UniswapV2State {
+    /// Validate that the reserves in the state match the on-chain token balances
+    async fn validate(
+        &self,
+        provider: Arc<RootProvider<Ethereum>>,
+        block_id: u64,
+        component_id: &Bytes,
+        tokens: &[Token],
+    ) -> Result<bool, Box<dyn std::error::Error + Send + Sync>> {
+        if tokens.len() != 2 {
+            return Err("UniswapV2 pool must have 2 tokens".into());
+        }
+
+        let pool_address = Address::from_slice(&component_id[..20]);
+        let token_0 = &tokens[0];
+        let token_1 = &tokens[1];
+
+        let token_0_address = Address::from_slice(&token_0.address[..20]);
+        let token_1_address = Address::from_slice(&token_1.address[..20]);
+
+        let token_0_contract = IERC20::new(token_0_address, provider.clone());
+        let token_1_contract = IERC20::new(token_1_address, provider.clone());
+
+        // Get on-chain balances at the specified block
+        let balance_0 = token_0_contract
+            .balanceOf(pool_address)
+            .block(block_id.into())
+            .call()
+            .await?;
+
+        let balance_1 = token_1_contract
+            .balanceOf(pool_address)
+            .block(block_id.into())
+            .call()
+            .await?;
+
```

Comment:
oh I think this will be very slow.. 🤔 Ideally we would need to batch all the node requests together with `batch.add_call(..)`. that's why I said on slack that maybe this would have to be divided into two?

> Maybe this needs to be split into two parts?
> - One for the RPC calls for the correct balances for example
> - One to compare the values/read the return values?
> - OR the return could be a mix of both: the contract and methods to call and then what to compare the results with.